### PR TITLE
[virt-launcher]: Perform graceful shutdown though ACPI without depending on virt-handler

### PR DESCRIPTION
**Pull request by iholder-redhat**
**What this PR does / why we need it**:
Currently, when a signal such as SIGTERM arrives to virt-launcher, the following happens:
* virt-launcher edits its domain metadata by calling `MarkGracefulShutdownVMI()`
* VMController (which is part of virt-handler package) identifies that as part of its control loop and calls virt-lacunher's `SignalShutdownVMI()`
* virt-launcher then shuts down the domain with ACPI.

After this PR, when a signal such as SIGTERM arrives to virt-launcher, it will skip informing virt-handler and take care to shut down itself. This will make graceful shutdown mechanism more resilient and reliable, especially if virt-handler is not responsive.

One example for a scenario this might be very helpful is on a host shutdown (without draining), on which virt-handler and virt-launcher will get SIGTERM at the same time. On this scenario virt-launcher cannot count on virt-handler to shut it down.

**Release note**:

```release-note
Improve resiliency and reliability of virt-launcher graceful shutdowns
```

### CODE REVIEW

**Comment by iholder-redhat** [comment]
@vladikr

**Review by davidvossel — COMMENTED** [review]
Here are a few things to consider about this change.

Having the shutdown logic managed by virt-handler currently allows for both visibility (events are fired when graceful shutdown is starting) and eventual consistency (shutdown is continually tried by reconcile loop). In addition to this, it centralizes the shutdown procedure to a single component, which leaves less questions from a debug standpoint about what component is controlling what flow.

> One example for a scenario this might be very helpful is on a host shutdown (without draining), on which virt-handler and virt-launcher will get SIGTERM at the same time. On this scenario virt-launcher cannot count on virt-handler to shut it down.

If virt-handler can shutdown at the exact same time as virt-launcher, we have bigger issues here than ACPI. The VMI wouldn't be able to migrate during the node drain. Daemonsets are typically drained last.


> it will skip informing virt-handler and take care to shut down itself. 

virt-handler will still attempt the shutdown due to the vmi.DeletionTimestamp != nil. So it will cause a race between virt-launcher and virt-handler to shutdown the VMI. I'm not completely sure what happens if both virt-handler and virt-launcher race to call shutdown. It might cause some unexpected noise. For example, errors commonly occurring due to virt-handler attempting to shutdown and already shutdown domain.

Is this PR a result of an issue that's occurring now? Understanding how this manifests itself in the wild in a negative way might help the discussion a bit.

**Comment by vladikr** [comment]
> Here are a few things to consider about this change.
> 
> Having the shutdown logic managed by virt-handler currently allows for both visibility (events are fired when graceful shutdown is starting) and eventual consistency (shutdown is continually tried by reconcile loop). In addition to this, it centralizes the shutdown procedure to a single component, which leaves less questions from a debug standpoint about what component is controlling what flow.

Well, I don't think we have a single shutdown logic in general. The VMI can shutdown independently, just by a guest OS user shutdown, getting a SIGTERM from a node-level component / privileged user, or even a k8s user deleting a vmipod. 
I think it has value if virt-launcher, when getting a SIGTERM, will attempt to send an ACPI shutdown without relying on virt-handler, which may not be running for some reason.
Regardless of whether the virt-handler has managed to send an ACPI shutdown or not, the virt-launcher will still send a SIGTERM to QEMU, using the finalShutdownCallback [1].

Eventual consistency can be achieved in this case as well as the virt-launcher already has a `finalShutdownCallback` that is being called after `gracePeriodSeconds`. 


> 
> > One example for a scenario this might be very helpful is on a host shutdown (without draining), on which virt-handler and virt-launcher will get SIGTERM at the same time. On this scenario virt-launcher cannot count on virt-handler to shut it down.
> 
> If virt-handler can shutdown at the exact same time as virt-launcher, we have bigger issues here than ACPI. The VMI wouldn't be able to migrate during the node drain. Daemonsets are typically drained last.
> 
> > it will skip informing virt-handler and take care to shut down itself.

Perhaps we already have a race between the virt-handler and the virt-launcher shutting down.

> 
> virt-handler will still attempt the shutdown due to the vmi.DeletionTimestamp != nil. So it will cause a race between virt-launcher and virt-handler to shutdown the VMI. I'm not completely sure what happens if both virt-handler and virt-launcher race to call shutdown. It might cause some unexpected noise. For example, errors commonly occurring due to virt-handler attempting to shutdown and already shutdown domain.

I think we can coordinate it by setting a shutdown flag in virt-launcher to prevent this.
I agree that we need better coordination between the virt-launcher and the virt-handler.

> 
> Is this PR a result of an issue that's occurring now? Understanding how this manifests itself in the wild in a negative way might help the discussion a bit.

I think this is a result of a direct node shutdown. (@iholder-redhat knows the details better) In this case, all processes (including child processes) will get a SIGTERM.
So I'm not sure that this PR will necessarily help us in this case, but I'm in favor of having virt-launcher attempt to send ACPI shutdown directly and not relying on the virt-handler. 

[1] https://github.com/kubevirt/kubevirt/blob/main/cmd/virt-launcher/virt-launcher.go#L467

**Comment by acardace** [comment]
> Here are a few things to consider about this change.
> 
> Having the shutdown logic managed by virt-handler currently allows for both visibility (events are fired when graceful shutdown is starting) and eventual consistency (shutdown is continually tried by reconcile loop). In addition to this, it centralizes the shutdown procedure to a single component, which leaves less questions from a debug standpoint about what component is controlling what flow.

I think that the biggest plus of the current logic is that a VMI shutdown is part of virt-handler's sync-loop and I'd keep it like that to make sure we don't miss anything (even if the shutdown itself is running on the same node and communication happens on a unix socket).

> 
> > One example for a scenario this might be very helpful is on a host shutdown (without draining), on which virt-handler and virt-launcher will get SIGTERM at the same time. On this scenario virt-launcher cannot count on virt-handler to shut it down.
> 
> If virt-handler can shutdown at the exact same time as virt-launcher, we have bigger issues here than ACPI. The VMI wouldn't be able to migrate during the node drain. Daemonsets are typically drained last.

I don't think node draining comes into play here, when a node is drained we protect the VMI with a PDB and we migrate it to another node, no SIGTERM or any other signal is delivered to virt-launcher. (right?)
> 
> > it will skip informing virt-handler and take care to shut down itself.
> 
> virt-handler will still attempt the shutdown due to the vmi.DeletionTimestamp != nil. So it will cause a race between virt-launcher and virt-handler to shutdown the VMI. I'm not completely sure what happens if both virt-handler and virt-launcher race to call shutdown. It might cause some unexpected noise. For example, errors commonly occurring due to virt-handler attempting to shutdown and already shutdown domain.
> 
> Is this PR a result of an issue that's occurring now? Understanding how this manifests itself in the wild in a negative way might help the discussion a bit.

I believe both sides have good points (removing all logic from virt-handler or not), what I'd suggest instead is to keep the shutdown logic as is and just have a special case for when virt-handler receives a SIGTERM only on that occasion I'd have virt-handler trying to shutdown using ACPI as it means something at the OS-level is happening and we just cannot fight the init system if it decided to deliver SIGTERM to everyone as a result of the host shutting-down, this special-casing shouldn't interfere with any other logic.

**Comment by iholder-redhat** [comment]
Some clarifications:

- I got this idea from a [similar bug](https://bugzilla.redhat.com/show_bug.cgi?id=2006571) that deals with direct node shutdown (shutting down a node without draining it properly first). This PR does not resolve the bug and does not address its root issue - but might be helpful for similar situations. I think @vladikr phrased it perfectly: `I think it has value if virt-launcher, when getting a SIGTERM, will attempt to send an ACPI shutdown without relying on virt-handler, which may not be running for some reason.`
- It's not that all of the shutdown logic is removed from virt-handler now. The only responsibility that is transferred to virt-launcher is shutting itself down when receiving a deadly signal such as SIGTERM. virt-handler still manages all other shutdown reasons and virt-launcher won't ever shut itself down unless a SIGTERM is received.
-  As @acardace put it: `what I'd suggest instead is to keep the shutdown logic as is and just have a special case for when virt-handler receives a SIGTERM` (I assume he meant virt-launcher there). That's exactly right. All logic remains the same, but receiving a SIGTERM is indeed a special case since we have to shut down as quickly as we can under the assumption that every process in the OS got SIGTERM at the same time as us or even before. In such a scenario the current logic would work only under the assumption that virt-handler is available and responsive, that it would be quick enough to shut all virt-launchers down in time, while we do not control OS' grace periods. Many side-effects can arise in such an extreme case (networking might not work since services also got SIGTERM, etc').

Also, I fully agree with @davidvossel's point - a single component that manages everything and can monitor and alert is great, both for simplicity and visibility. But all of this doesn't apply on a direct OS shutdown where everyone gets SIGTERM and there's an urgent need to shut down ASAP.

**Comment by davidvossel** [comment]
> It's not that all of the shutdown logic is removed from virt-handler now. The only responsibility that is transferred to virt-launcher is shutting itself down when receiving a deadly signal such as SIGTERM. virt-handler still manages all other shutdown reasons and virt-launcher won't ever shut itself down unless a SIGTERM is received.

what shutdown procedure is virt-handler responsible for that doesn't also result in virt-launcher receiving SIGTERM?

I think that unless a guest is internally shutdown, all the external ways of shutting down a VMI involve marking the VMI for deletion. By marking a VMI for deletion (deletiontimestamp!=nil), the pod is also marked for deletion which in turn will result in SIGTERM being sent to virt-launcher.

If we shift the responsibility of graceful shutdown to virt-launcher, i'd like virt-launcher to be the only component attempting to coordinate the shutdown. In this PR, I believe both virt-handler's reconcile loop and virt-launcher's gracefulSHutdownCallback() are both trying to do the same thing concurrently.

**Comment by iholder-redhat** [comment]
> > It's not that all of the shutdown logic is removed from virt-handler now. The only responsibility that is transferred to virt-launcher is shutting itself down when receiving a deadly signal such as SIGTERM. virt-handler still manages all other shutdown reasons and virt-launcher won't ever shut itself down unless a SIGTERM is received.
> 
> what shutdown procedure is virt-handler responsible for that doesn't also result in virt-launcher receiving SIGTERM?

All of them :smiley: 

Eventually, every shutdown that virt-handler initiates involves calling to `ShutdownVirtualMachine()`, and after the grace period is over, `KillVirtualMachine()`. These are virt-launcher's client functions. This means that virt-handler **doesn't directly send SIGTERM** to virt-launcher nor directly doing any operation to shut down - but instead it asks virt-launcher to perform the graceful shutdown for itself.

This PR does **not** change any of this.

> I think that unless a guest is internally shutdown, all the external ways of shutting down a VMI involve marking the VMI for deletion. By marking a VMI for deletion (deletiontimestamp!=nil), the pod is also marked for deletion which in turn will result in SIGTERM being sent to virt-launcher.

Further to what I've written above, the only this does PR changes is the scenario in which virt-launcher is directly receiving SIGTERM. This **can't be caused** by virt-handler trying to shut down virt-launcher as I wrote above.

This can be caused though by a sudden node shutdown without draining. In such a scenario the OS can trigger a direct SIGTERM to virt-launcher. This is what this PR deals with.

In such a scenario, today, virt-launcher marks itself as needing shutdown. Then, virt-handler sees this mark, and uses its regular mechanism to shutdown virt-launcher, which is calling virt-launcher's `ShutdownVirtualMachine()` and (if grace period expires) `KillVirtualMachine()` as described above.

To sum it all up, when the OS sends SIGTERM directly to virt-launcher, this happens:
- virt-launcher gets SIGTERM and marks itself for deletion
- virt-handler sees the mark and calls virt-launcher's `ShutdownVirtualMachine()`
- virt-launcher shuts itself down gracefully

In the specific scenario in which virt-launcher gets a direct SIGTERM from the OS, steps (1) and (2) are not only unnecessary, but they are dangerous since virt-handler might not be responsive, or response too slowly, therefore we're jeopardizing the ability of virt-launcher to terminate itself without a good reason.

> If we shift the responsibility of graceful shutdown to virt-launcher, i'd like virt-launcher to be the only component attempting to coordinate the shutdown. In this PR, I believe both virt-handler's reconcile loop and virt-launcher's gracefulSHutdownCallback() are both trying to do the same thing concurrently.

Today:
virt-handler has the responsibility to trigger shutdown
virt-launcher has the responsibility of performing the shutdown.

With this PR's changes:
Only when a direct SIGTERM is sent from OS - virt-launcher would have the responsibility of performing the shutdown without waiting for virt-handler to trigger it.

**Comment by davidvossel** [comment]
> Eventually, every shutdown that virt-handler initiates involves calling to ShutdownVirtualMachine(), and after the grace period is over, KillVirtualMachine(). These are virt-launcher's client functions. This means that virt-handler doesn't directly send SIGTERM to virt-launcher nor directly doing any operation to shut down - but instead it asks virt-launcher to perform the graceful shutdown for itself.

The kubelet sends SIGTERM to virt-launcher every time a VMI is deleted because we're deleting the pod in virt-controller when we see the VMI is being deleted.

That's why I asked the question "what shutdown procedure is virt-handler responsible for that doesn't also result in virt-launcher receiving SIGTERM?". I believe the answer is none of them due to the pod deletion + kubelet tear down. 

>Further to what I've written above, the only this does PR changes is the scenario in which virt-launcher is directly receiving SIGTERM. This can't be caused by virt-handler trying to shut down virt-launcher as I wrote above.

but kubelet is shutting down virt-launcher.

Sequence of events.

1. delete VMI
2. virt-controller deletes vmi pod
3. virt-handler sees vmi is being deleted and starts shutdown
4. virt-launcher receives and traps SIG_TERM sent by kubelet.
5. eventually vmi pod terminates.

Right now, items 3 and 4 get routed through virt-handler due to virt-launcher only trapping the SIG_TERM and alerting virt-handler that a graceful shutdown needs to be processed.

I think what we have in your PR is both virt-handler and virt-launcher are both converging on libvirt to shutdown the domain at the same time.

> With this PR's changes:
Only when a direct SIGTERM is sent from OS - virt-launcher would have the responsibility of performing the shutdown without waiting for virt-handler to trigger it.

my concern is that with your PR, virt-handler still signals virt-launcher to do the shutdown under nearly every ordinary shutdown sequence we have today where the VMI is shutdown by deleting the VMI object.

**Comment by iholder-redhat** [comment]
> The kubelet sends SIGTERM to virt-launcher every time a VMI is deleted because we're deleting the pod in virt-controller when we see the VMI is being deleted.

That's a good point, I didn't realize that until now. Thanks for educating me here.

That makes me fully agree with what you wrote earlier:
> If we shift the responsibility of graceful shutdown to virt-launcher, i'd like virt-launcher to be the only component attempting to coordinate the shutdown.

Looking at the code again I see that the only case in which virt-handler deals with shutting down is, as you said, the case in which the VMI is being deleted. 

If VMI is being deleted, virt-handler doesn't need to do anything. SIGTERM will be sent to virt-launcher which will handle the shutdown.

I will add a commit that would remove virt-handler shut down logic entirely.

File: pkg/virt-handler/cmd-client/client.go [hunk 1]
```diff
@@ -448,14 +446,6 @@ func (c *VirtLauncherClient) SoftRebootVirtualMachine(vmi *v1.VirtualMachineInst
...
 }
 
-func (c *VirtLauncherClient) ShutdownVirtualMachine(vmi *v1.VirtualMachineInstance) error {
-	return c.genericSendVMICmd("Shutdown", c.v1client.ShutdownVirtualMachine, vmi, &cmdv1.VirtualMachineOptions{})
-}
-
-func (c *VirtLauncherClient) KillVirtualMachine(vmi *v1.VirtualMachineInstance) error {
-	return c.genericSendVMICmd("Kill", c.v1client.KillVirtualMachine, vmi, &cmdv1.VirtualMachineOptions{})
-}
-
```
Comment (davidvossel) [inline pkg/virt-handler/cmd-client/client.go:458]: This kind of sucks, but I think we have to keep this around.

Here's the issue. Updates... Think about this scenario.

- Kubevirt v0.48 is deployed
- vmi A is running with virt-launcher v0.48
- KubeVirt is updated to v0.49

at this point, shutting down vmi A properly requires the old virt-handler logic, but that old virt-handler logic won't fire if we strip out the api entirely for 0.49.


Here's a potential solution. Keep the api, but simply make the new virt-launcher no-op the shutdown when the shutdown is already in progress. That way old virt-laucher's still have shutdown called, and the new ones just handle that shutdown command differently.

**Review by davidvossel — COMMENTED** [review]
> If VMI is being deleted, virt-handler doesn't need to do anything. SIGTERM will be sent to virt-launcher which will handle the shutdown.
> I will add a commit that would remove virt-handler shut down logic entirely.

nice, this is ultimately what I was hoping for. checkout my comment in line about handling updates.

File: pkg/virt-handler/cmd-client/client.go (hunk 1 above)
Comment (iholder-redhat) [reply pkg/virt-handler/cmd-client/client.go:458]: Good point, thanks for raising that. I'm on it.

Perhaps it's reasonable to deprecate it entirely in the future?
How do we keep track of such need-to-be-deprecated code?

Comment (davidvossel) [reply pkg/virt-handler/cmd-client/client.go:458]: we don't have a great deprecation policy right now unfortunately :/

**Comment by iholder-redhat** [comment]
Hey @davidvossel,

After digging deeper into the code I've come to realize this is more complicated that I thought. The main challenge here is updating the grace period.

Currently, since the handler is solely responsible for shutdowns, it:
* Gets the most updated grace period (directly from the VMI's spec)
* Calculates whether the grace period exceeded
* Triggers `SignalShutdownVMI()` as long as grace period didn't exceed, `KillVMI()` otherwise.

The problem with moving the logic into virt-handler's hands is that virt-launcher can't really know the updated grace period since it can't talk directly to the API server and ask for a VMI object. This basically means that the VMI's grace period field would become meaningless.

Consider this example:
```shell
> k create -f my-vmi.yaml # In this yaml file, a grace period of 100 seconds is defined
...
> k edit vmi my-vmi # The vmi's grace period is changed to 50 seconds
...
> k delete -f my-vmi.yaml
```

Currently the `delete` operation would trigger a SIGTERM to virt-launcher. virt-launcher would "mark itself for shutdown", virt-handler would notice that and will manage shutdown logic as I described above. In other words - the new defined grace period will take effect. But if logic will move to virt-launcher - then during a SIGTERM it won't have any idea what is the currently updated grace period.

One possible solution is to, somehow, get notified if grace period is updated and tell virt-launcher (through gRPC) that grace period should be updated. Not sure how or if that's possible / reasonable.

Another options is to disallow the update of grace period entirely (or to set it back during reconciliation loop). This might (?) be a reasonable approach, except we have `virtctl` which needs to have this capability, especially for `stop --force --grace-period=0`. Currently virtctl simply updates grace period for the VMI and then deletes it - similarly to the shell example above. To overcome this virtctl would have to ask launcher client to shut down the VMI (though `SignalShutdownVMI()` or other similar gRPC call) and ask for a new grace period directly from virt-launcher and only then delete the VMI. While this approach is possible it will make code more error prone and harder to understand, since the responsibility is moved to whoever tries to shutdown a VMI which will need to be aware of all of this.

Considering the above I'm not sure what is the right path forward here. When opening this PR my assumption was that SIGTERM for virt-launcher is a rare event that can be dealt with separately, but this is not the case as I understand now.

WDYT?

@vladikr would appreciate your thought on this as well.

**Comment by davidvossel** [comment]
> Considering the above I'm not sure what is the right path forward here. When opening this PR my assumption was that SIGTERM for virt-launcher is a rare event that can be dealt with separately, but this is not the case as I understand now.

Should we consider the option of "proceed as is"? and accept that if someone issues a `shutdown` locally on a node that their VMs likely will not receive a ACPI shutdown?

I believe the situation we're talking about solving here is graceful shutdown of VMs when virt-handler isn't a part of the picture. This would mean termination is occurring outside of the k8s ecosystem at the host level. Do pods even have their grace periods observed in the scenario we're solving for?

It would be nice for virt-launcher to be able to coordinate the shutdown, but if this is getting overly complicated it's unclear if the benefits outweigh the risks and complexity involved.

**Comment by iholder-redhat** [comment]
> Should we consider the option of "proceed as is"? and accept that if someone issues a `shutdown` locally on a node that their VMs likely will not receive a ACPI shutdown?
> 
> I believe the situation we're talking about solving here is graceful shutdown of VMs when virt-handler isn't a part of the picture. This would mean termination is occurring outside of the k8s ecosystem at the host level. Do pods even have their grace periods observed in the scenario we're solving for?
> 
> It would be nice for virt-launcher to be able to coordinate the shutdown, but if this is getting overly complicated it's unclear if the benefits outweigh the risks and complexity involved.

Yup.. a bit disappointing but I tend to agree.
Thanks a lot for exploring this with me.

/close

**Comment by davidvossel** [comment]
> Yup.. a bit disappointing but I tend to agree.
> Thanks a lot for exploring this with me.

@iholder-redhat lets consider this idea on hold for now until we have a use case that forces us to re-evaluate our position. Right now I don't think the risk/reward adds up in favor of us making the change, but I'm certainly open to the idea if new info comes into the picture.

**Pull request closed without merging**